pkg/messaging/mqtt/pubsub.go

```diff
@@ -87,18 +88,21 @@ func (ps pubsub) Subscribe(id, topic string, handler messaging.MessageHandler) e
 			if err != nil {
 				return err
 			}
-			if len(ps.subscriptions) == 0 {
+			// ps.subscriptions[id] was updated after call to Unsubscribe(), but s is a copy, so it wasn't updated
```

Comment:
@AryanGodara @arvindh123 A combination of your ideas can be something like this:

1. First, this entire `switch` can be replaced with:
```go
	// If the client exists, check if it's subscribed to the topic and unsubscribe if needed.
	switch ok {
	case true:
		if ok := s.contains(topic); ok {
			if err := s.unsubscribe(topic, ps.timeout); err != nil {
				return err
			}
		}
	default:
		client, err := newClient(ps.address, id, ps.timeout)
		if err != nil {
			return err
		}
		s = subscription{
			client: client,
			topics: []string{},
		}
	}
        s.topics = append(s.topics, topic)
	ps.subscriptions[id] = s
...
```
2. Where unsubscribe looks like:
```go
func (s *subscription) unsubscribe(topic string, timeout time.Duration) error {
	token := s.client.Unsubscribe(topic)
	if token.Error() != nil {
		return token.Error()
	}

	if ok := token.WaitTimeout(timeout); !ok {
		return ErrUnsubscribeTimeout
	}
	if ok := s.delete(topic); !ok {
		return ErrUnsubscribeDeleteTopic
	}
	return nil
}
```
3. Modify `Unsubscribe` to also utilize `s.unsubscribe(topic, timeout)` accordingly.

Reply:
I've implemented this, and all tests are passing!